File: python/dmarebrandsbot/cogs/account.py
from __future__ import annotations

import discord
from discord import app_commands
from discord.ext import commands

from ..formatting import BRAND, embed, iso_stamp, money
from ..permissions import describe, requires, scopes_for


class Account(commands.Cog):
    def __init__(self, bot: commands.Bot) -> None:
        self.bot = bot

    @app_commands.command(description="Show the partner account this bot is connected to")
    @requires("account.read")
    async def account(self, interaction: discord.Interaction) -> None:
        me = await self.bot.partner_api.me()
        brand = me.get("brand") or {}
        stats = me.get("stats") or {}

        view = embed(f"Partner account: {me.get('username')}", BRAND)
        view.description = (
            f"Trading as **{brand['name']}**" if brand.get("name") else "No brand name set yet."
        )
        view.add_field(name="Balance", value=money(me.get("balance_usd")), inline=True)
        view.add_field(name="Discount", value=f"{me.get('discount_pct')}%", inline=True)
        view.add_field(name="Customers", value=str(stats.get("customers", 0)), inline=True)
        view.add_field(name="Keys in stock", value=str(stats.get("keys_unused", 0)), inline=True)
        view.add_field(name="Keys active", value=str(stats.get("keys_active", 0)), inline=True)
        view.add_field(name="Spent all time", value=money(stats.get("spent_usd")), inline=True)

        if brand.get("store_url"):
            view.url = brand["store_url"]

        await interaction.edit_original_response(embed=view)

    @app_commands.command(description="List every plan you can buy, with your discounted price")
    @requires("account.read")
    async def plans(self, interaction: discord.Interaction) -> None:
        result = await self.bot.partner_api.plans()
        rows = result.get("data") or []

        if not rows:
            await interaction.edit_original_response(content="No plans are available right now.")
            return

        view = embed("Plans", BRAND)
        view.description = "\n\n".join(
            f"**{plan['label']}** `{plan['id']}`\n"
            f"{plan['days']} days · {money(plan['your_price_usd'])} for you · "
            f"list {money(plan['list_price_usd'])} · you save {money(plan['saving_usd'])}"
            for plan in rows
        )
        await interaction.edit_original_response(embed=view)

    @app_commands.command(description="Headline numbers for the partner account")
    @requires("account.read")
    async def stats(self, interaction: discord.Interaction) -> None:
        s = await self.bot.partner_api.stats()
        view = embed("Stats", BRAND)
        view.add_field(name="Keys issued", value=str(s.get("keys_total", 0)), inline=True)
        view.add_field(name="In stock", value=str(s.get("keys_unused", 0)), inline=True)
        view.add_field(name="Active", value=str(s.get("keys_active", 0)), inline=True)
        view.add_field(name="Expired", value=str(s.get("keys_expired", 0)), inline=True)
        view.add_field(name="Customers", value=str(s.get("customers", 0)), inline=True)
        view.add_field(name="Balance", value=money(s.get("balance_usd")), inline=True)
        view.add_field(name="Spent all time", value=money(s.get("spent_usd")), inline=True)
        view.add_field(name="HWID resets (30d)", value=str(s.get("hwid_resets_30d", 0)), inline=True)
        await interaction.edit_original_response(embed=view)

    @app_commands.command(description="Show your white-label domains and whether they are live")
    @requires("domains.read")
    async def domains(self, interaction: discord.Interaction) -> None:
        result = await self.bot.partner_api.domains()
        rows = result.get("data") or []
        target = result.get("cname_target")

        if not rows:
            await interaction.edit_original_response(
                content=(
                    "No domains set up yet. Add one at "
                    "https://partners.dmarebrands.st/partners/branding\n"
                    f"Point a CNAME at `{target or 'the target shown in the panel'}`."
                )
            )
            return

        lines = []
        for domain in rows:
            state = (
                "Live"
                if domain.get("live")
                else f"Not live yet ({domain.get('status')} / ssl {domain.get('ssl_status')})"
            )
            checked = f" · checked {iso_stamp(domain['checked_at'])}" if domain.get("checked_at") else ""
            error = f"\n{domain['last_error']}" if domain.get("last_error") else ""
            lines.append(f"**{domain['hostname']}** · {domain['kind']}\n{state}{checked}{error}")

        view = embed("Your domains", BRAND)
        view.description = "\n\n".join(lines)
        if target:
            view.set_footer(text=f"CNAME target: {target}")

        await interaction.edit_original_response(embed=view)

    @app_commands.command(description="Show what this bot will let you do")
    async def whoami(self, interaction: discord.Interaction) -> None:
        granted = scopes_for(interaction.user, self.bot.partner_config.role_scopes)
        view = embed("Your access", BRAND)
        view.description = (
            f"You have these scopes:\n`{describe(granted)}`"
            if granted
            else "None of your roles are configured for this bot, so every command will be refused."
        )
        await interaction.edit_original_response(embed=view)


async def setup(bot: commands.Bot) -> None:
    await bot.add_cog(Account(bot))

File: node/src/commands/account.js
import { SlashCommandBuilder } from "discord.js";
import { BRAND, embed, isoStamp, money } from "../format.js";
import { describeScopes } from "../permissions.js";

export const account = {
  data: new SlashCommandBuilder()
    .setName("account")
    .setDescription("Show the partner account this bot is connected to"),
  scopes: "account.read",
  async run({ interaction, api }) {
    const me = await api.me();
    const view = embed(`Partner account: ${me.username}`, BRAND)
      .setDescription(me.brand?.name ? `Trading as **${me.brand.name}**` : "No brand name set yet.")
      .addFields(
        { name: "Balance", value: money(me.balance_usd), inline: true },
        { name: "Discount", value: `${me.discount_pct}%`, inline: true },
        { name: "Customers", value: String(me.stats?.customers ?? 0), inline: true },
        { name: "Keys in stock", value: String(me.stats?.keys_unused ?? 0), inline: true },
        { name: "Keys active", value: String(me.stats?.keys_active ?? 0), inline: true },
        { name: "Spent all time", value: money(me.stats?.spent_usd), inline: true }
      );

    if (me.brand?.store_url) view.setURL(me.brand.store_url);
    await interaction.editReply({ embeds: [view] });
  },
};

export const plans = {
  data: new SlashCommandBuilder()
    .setName("plans")
    .setDescription("List every plan you can buy, with your discounted price"),
  scopes: "account.read",
  async run({ interaction, api }) {
    const result = await api.plans();
    const rows = result.data ?? [];

    if (!rows.length) {
      await interaction.editReply({ content: "No plans are available right now." });
      return;
    }

    const view = embed("Plans", BRAND).setDescription(
      rows
        .map(
          (p) =>
            `**${p.product} · ${p.label}** \`${p.id}\`\n${p.days} days · ${money(p.your_price_usd)} for you · list ${money(
              p.list_price_usd
            )} · you save ${money(p.saving_usd)}`
        )
        .join("\n\n")
    );

    await interaction.editReply({ embeds: [view] });
  },
};

export const stats = {
  data: new SlashCommandBuilder()
    .setName("stats")
    .setDescription("Headline numbers for the partner account"),
  scopes: "account.read",
  async run({ interaction, api }) {
    const s = await api.stats();
    const view = embed("Stats", BRAND).addFields(
      { name: "Keys issued", value: String(s.keys_total), inline: true },
      { name: "In stock", value: String(s.keys_unused), inline: true },
      { name: "Active", value: String(s.keys_active), inline: true },
      { name: "Expired", value: String(s.keys_expired), inline: true },
      { name: "Customers", value: String(s.customers), inline: true },
      { name: "Balance", value: money(s.balance_usd), inline: true },
      { name: "Spent all time", value: money(s.spent_usd), inline: true },
      { name: "HWID resets (30d)", value: String(s.hwid_resets_30d), inline: true }
    );
    await interaction.editReply({ embeds: [view] });
  },
};

export const domains = {
  data: new SlashCommandBuilder()
    .setName("domains")
    .setDescription("Show your white-label domains and whether they are live"),
  scopes: "domains.read",
  async run({ interaction, api }) {
    const result = await api.domains();
    const rows = result.data ?? [];

    if (!rows.length) {
      await interaction.editReply({
        content: `No domains set up yet. Add one at https://partners.dmarebrands.st/partners/branding\nPoint a CNAME at \`${result.cname_target ?? "the target shown in the panel"}\`.`,
      });
      return;
    }

    const view = embed("Your domains", BRAND).setDescription(
      rows
        .map((d) => {
          const state = d.live ? "Live" : `Not live yet (${d.status} / ssl ${d.ssl_status})`;
          const checked = d.checked_at ? ` · checked ${isoStamp(d.checked_at)}` : "";
          const error = d.last_error ? `\n${d.last_error}` : "";
          return `**${d.hostname}** · ${d.kind}\n${state}${checked}${error}`;
        })
        .join("\n\n")
    );

    if (result.cname_target) view.setFooter({ text: `CNAME target: ${result.cname_target}` });
    await interaction.editReply({ embeds: [view] });
  },
};

export const whoami = {
  data: new SlashCommandBuilder()
    .setName("whoami")
    .setDescription("Show what this bot will let you do"),
  scopes: null,
  async run({ interaction, granted }) {
    const view = embed("Your access", BRAND).setDescription(
      granted.size
        ? `You have these scopes:\n\`${describeScopes(granted)}\``
        : "None of your roles are configured for this bot, so every command will be refused."
    );
    await interaction.editReply({ embeds: [view] });
  },
};
